Expand DEVELOPMENT proxy section (#192)
radicle-dev/radicle-upstream: DEVELOPMENT.md

git submodule foreach "git fetch --all"
```

Then run tests as usual: `cargo test --all-features --all-targets`.

> We strive for two kind of tests, the classic unit tests inside of the

Reviewer: This sentence doesn't read well.

```suggestion
We strive for two kinds of tests:

1. The classic unit tests, located in the respective implementation modules
2. The integration tests, focused on asserting the API that the proxy provides from a high-level consumer perspective. These can be found under `proxy/tests`.

To find out where to place and how to layout tests check Rust book [test chapter][rt].
```

implementation files. Additionally we have a suit of integration tests esp. to
assert the API the proxy provides from a high-level consumer perspective, these
can be found under `proxy/tests`. To find out where to place and how to layout

> tests check Rust book [test chapter][rt].

Reviewer 2: ```suggestion
We strive for two kinds of tests: classic unit tests contained in implementation
files and integration tests. The integration tests are meant to assert correctness
of the API provided by the proxy, these can be found under `proxy/tests`. To find
out where to place and how to lay out tests, check the Rust book [test chapter][rt].
```

Reviewer: This is also good!

### File structure

> We understand the GraphQL API as one potential surface to expose the

Reviewer: > We understand the GraphQL API as one potential surface to expose the Application domain logic

I am not sure I understand this. Could you explain what you mean? How could we explain this in a simpler way?

Reviewer 3: Maybe:

> The GraphQL API is one potential surface for exposing the application's domain logic.

But I don't understand what "potential" is doing in this sentence. Since we already use it, isn't it "actual"? If so, it could be dropped and the sentence rephrased:

> The GraphQL API exposes the application's domain logic. Therefore we try to treat it as a thin layer...

> Application domain logic. Therefore we try to treat it as thin layer exposing

Reviewer 2: ```suggestion
Application domain logic. Therefore we try to treat it as a thin layer exposing
```

> well-typed entities. The heavy lifting is done in the modules named after the

Reviewer: > The heavy lifting

I'd rather use more specific language here, especially since I do weightlifting :P

Reviewer 3: Perhaps:

> Complementing this thin layer are the related module `CoCo` (link) which implements code collaboration features, and the `Registry` module which implements interactions with the Radicle registry blockchain. By isolating concerns this way...

> protocols we consume - `CoCo` for code collaboration and `Registry` for global

Reviewer 4: Links to the two repos could be helpful here for an outside reader.

Author: Expanded in 536b402.

> unique entries for users, projects and organistions. By isolating concerns in

Reviewer 2: ```suggestion
unique entries for users, projects and organisations. By isolating concerns
```

> this weay, we hope to enable ease-of-contribution to downstream teams.

Reviewer 2: ```suggestion
this way, we hope to enable ease-of-contribution to downstream teams.
```

Empowering them to reflect changes in their public APIs easily with code
contributions to Upstream.

```
proxy/src/
├── coco.rs
├── env.rs
├── error.rs
├── graphql
│   ├── api.rs
│   ├── error.rs
│   ├── project.rs
│   └── schema.rs
├── graphql.rs
├── lib.rs
├── main.rs
└── registry.rs
```

💡 *You'll have to run the submodule commands every time there are any updates

> to the test fixture repository.*

Reviewer 2: This should be moved back to the `testing` section below the `git submodule` commands (around L:408).

## Scripts

> To get a list of all available script commands, run: `yarn run`.

Reviewer: These need to be run under `app`, right?

Reviewer 2: There is no app anymore, we moved it to the top-level (screens and components now live in `ui`).

Reviewer: ah, cool. 👍

Here is a list of the most commonly used ones:

```sh
yarn start            # Start Upstream in development mode

yarn test             # Run Cypress end-to-end tests
yarn test:debug       # Show the Cypress GUI, handy for visual debugging

yarn dist             # Bundles Upstream into an installable package

yarn generate:colors  # Update color CSS variables in public/colors.css from
                      # colors.json

yarn graphiql         # Open GraphiQL with explorer extension in browser

yarn release          # Start a two-step process to cut a new release, for more
                      # details have a look at ../DEVELOPMENT.md

yarn prettier:check   # Check UI code formatting
yarn prettier:write   # Auto-format UI code
yarn lint             # Check UI code for linting errors

> ```

Reviewer 2: We also have a bunch of helper scripts for the proxy in our `package.json`, should we maybe include those here too with proper descriptions?
```
proxy:build          # cd proxy && cargo build --all-features --all-targets
proxy:build:release  # cd proxy && cargo build --release --all-features --all-targets
proxy:build:test     # cd proxy && cargo test --no-run
proxy:start          # cd proxy && cargo run --release -- --registry=emulator
proxy:start:test     # cd proxy && cargo run -- --registry=emulator --test
```

Author: Generally I find these outputs of cli options to be prone to quickly get outdated. I'd rather show a command which shows all the commands.

Reviewer 2: I know what you mean. This section expands a bit on the commands however. It's meant to explain when they should be used. As the `yarn run` command can only show the commands themselves, but there is no way to inline documentation in `package.json`.

I just provided the `# cd proxy && cargo run` comments as a reference, the idea was that you'd write the explanation in a comment...

Author: Addressed in 4b037b7

git submodule foreach "git fetch --all"
```

Then run tests as usual: `cargo test --all-features --all-targets`.

We strive for two kinds of tests: classic unit tests contained in implementation
files and integration tests. The integration tests are meant to assert correctness
of the API provided by the proxy, these can be found under `proxy/tests`. To find
out where to place and how to lay out tests, check the Rust book [test chapter][rt].

> 💡 *You'll have to run the submodule commands every time there are any updates

Reviewer 2: This should be moved below the `git submodule` commands on L442.

to the test fixture repository.*

### File structure

The GraphQL API exposes the application's domain logic. Therefore we try to treat
it as a thin layer exposing well-typed entities. The heavy lifting is done in the
modules named after the protocols we consume - [radicle-link](https://github.com/radicle-dev/radicle-link),
through it [radicle-surf](https://github.com/radicle-dev/radicle-surf/), for code

> collaboration and [radicle-registry](https://github.com/radicle-dev/radicle-registry/)

Reviewer 2: Let's move the links to the end of the document.

├── graphql
│   ├── api.rs
│   ├── error.rs
│   ├── project.rs
│   └── schema.rs
├── graphql.rs
├── lib.rs
├── main.rs
└── registry.rs
```

## Scripts

To get a list of all available script commands, run: `yarn run`.
Here is a list of the most commonly used ones:

```sh
yarn start                # Start Upstream in development mode

yarn test                 # Run Cypress end-to-end tests
yarn test:debug           # Show the Cypress GUI, handy for visual debugging

yarn dist                 # Bundles Upstream into an installable package

yarn generate:colors      # Update color CSS variables in public/colors.css from
                          # colors.json

yarn graphiql             # Open GraphiQL with explorer extension in browser

yarn release              # Start a two-step process to cut a new release, for more
                          # details have a look at ../DEVELOPMENT.md

yarn prettier:check       # Check UI code formatting
yarn prettier:write       # Auto-format UI code
yarn lint                 # Check UI code for linting errors

yarn proxy:build          # Build the proxy binary
yarn proxy:build:release  # Build the release version of the proxy, stripped of

> debug symbols

Reviewer 2: ```suggestion
                          # debug symbols
```

yarn proxy:start          # Start only the proxy with its default configuration
yarn proxy:start:test     # Start the proxy in test mode, where state is

> isolated and lives in memory or temporary directories

Reviewer 2: ```suggestion
                          # isolated and lives in memory or temporary directories
```
